Realm/ObjectServerTests/RLMObjectServerTests.m

```diff
@@ -187,362 +321,279 @@ - (void)testUploadChangesWhenRealmOutOfScope {
-        [r addObject:[[SyncObject alloc] initWithValue:@[@"child-1"]]];
-        [r addObject:[[SyncObject alloc] initWithValue:@[@"child-2"]]];
-        [r commitWriteTransaction];
-        [user waitForUploadToFinish:url];
+        WAIT_FOR_DOWNLOAD(user, url);
+        [self addSyncObjectsToRealm:r descriptions:@[@"child-1", @"child-2"]];
+        WAIT_FOR_UPLOAD(user, url);
         CHECK_COUNT(3, SyncObject, r);
     }
 }
 
 /// A Realm that was opened while a user was logged out should be able to start uploading if the user logs back in.
 - (void)testLogBackInDeferredRealmUpload {
-    NSURL *url = [NSURL URLWithString:@"realm://localhost:9080/~/testBasicSync"];
+    NSURL *url = REALM_URL();
     // Log in the user.
-    RLMSyncUser *user = [self logInUserForCredential:[RLMObjectServerTests basicCredential:self.isParent]
+    RLMSyncUser *user = [self logInUserForCredential:[RLMObjectServerTests basicCredentialWithName:ACCOUNT_NAME()
+                                                                                     createAccount:self.isParent]
                                               server:[RLMObjectServerTests authServerURL]];
     NSError *error = nil;
     if (self.isParent) {
+        // Semaphore for knowing when the Realm is successfully opened for sync.
+        dispatch_semaphore_t sema = dispatch_semaphore_create(0);
         RLMRealmConfiguration *config = [RLMRealmConfiguration defaultConfiguration];
         config.syncConfiguration = [[RLMSyncConfiguration alloc] initWithUser:user realmURL:url];
         [user logOut];
         // Open a Realm after the user's been logged out.
+        [self primeSyncManagerWithSemaphore:sema];
         RLMRealm *r = [RLMRealm realmWithConfiguration:config error:&error];
         XCTAssertNil(error, @"Error when opening Realm: %@", error);
-        [r beginWriteTransaction];
-        [r addObject:[[SyncObject alloc] initWithValue:@[@"parent-1"]]];
-        [r commitWriteTransaction];
+        [self addSyncObjectsToRealm:r descriptions:@[@"parent-1"]];
         CHECK_COUNT(1, SyncObject, r);
-        user = [self logInUserForCredential:[RLMObjectServerTests basicCredential:NO]
+        user = [self logInUserForCredential:[RLMObjectServerTests basicCredentialWithName:ACCOUNT_NAME()
+                                                                            createAccount:NO]
                                      server:[RLMObjectServerTests authServerURL]];
-        // Wait for the sessions to asynchronously rebind
-        // FIXME: once new token system is in this will be unnecessary
-        usleep(500000);
-        [r beginWriteTransaction];
-        [r addObject:[[SyncObject alloc] initWithValue:@[@"parent-2"]]];
-        [r addObject:[[SyncObject alloc] initWithValue:@[@"parent-3"]]];
-        [r commitWriteTransaction];
-        [user waitForUploadToFinish:url];
+        // Wait for the Realm's session to be bound.
+        WAIT_FOR_SEMAPHORE(sema, 30);
```

Comment:
There's a real problem if this takes 30 seconds to happen. What's the lowest value that reliably passes?